patches/api/0374-Add-Recipe-API.patch

```diff
@@ -0,0 +1,273 @@
+--- a/src/main/java/org/bukkit/Server.java
++++ b/src/main/java/org/bukkit/Server.java
+@@ -2002,5 +2002,14 @@ public interface Server extends PluginMessageRecipient, net.kyori.adventure.audi
+      * @return the potion brewer
+      */
+     @NotNull org.bukkit.potion.PotionBrewer getPotionBrewer();
++
++    /**
++     * Gets all recipes of a certain type.
++     *
++     * @param recipeType recipe type to check against
++     * @return all recipes registered under the input type
++     */
++    @NotNull
++    <R extends org.bukkit.inventory.Recipe> List<R> getAllRecipes(io.papermc.paper.inventory.@NotNull RecipeType<R, ?> recipeType);
+     // Paper end
+ }
+diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
+index 8a688583e65cd22e0417f9fd24e51803486d095e..c16eb81d1491e68187f2140ad5a12a05c94ac2f1 100644
+--- a/src/main/java/org/bukkit/World.java
++++ b/src/main/java/org/bukkit/World.java
+@@ -3929,6 +3929,84 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
+     @Nullable
+     public DragonBattle getEnderDragonBattle();
+ 
++    // Paper start
++    /**
++     * Gets a recipe from the items inside of an inventory.
++     *
++     * @param recipeType recipe type to check against
++     * @param inventory the inventory to check
++     * @return a recipe resulting from the input items, or null if no result is found
++     */
++    @Nullable
++    <R extends org.bukkit.inventory.Recipe, I extends org.bukkit.inventory.Inventory> R getRecipe(io.papermc.paper.inventory.@NotNull RecipeType<R, I> recipeType, @NotNull I inventory);
++
++    /**
++     * Gets a recipe from a collection of items.
++     * <p>
++     * The collection should be ordered from left-to-right, top-to-bottom.
++     * The zeroth index would be the top left slot of a crafting table,
++     * the first index would be the top middle slot, et cetera.
++     *
++     * @param recipeType recipe type to check against
++     * @param items items to craft with
++     * @return a recipe resulting from the input items, or null if no result is found
++     */
++    @Nullable
++    <R extends org.bukkit.inventory.Recipe> R getRecipe(io.papermc.paper.inventory.@NotNull RecipeType<R, ?> recipeType, @NotNull Collection<@Nullable ItemStack> items);
++
++    /**
++     * Gets a recipe from an index of items.
++     * <p>
++     * The index should be ordered from left-to-right, top-to-bottom.
++     * The zeroth index would be the top left slot of a crafting table,
++     * the first index would be the top middle slot, et cetera.
++     *
++     * @param recipeType recipe type to check against
++     * @param items items to craft with
++     * @return a recipe resulting from the input items, or null if no result is found
++     */
++    @Nullable
++    <R extends org.bukkit.inventory.Recipe> R getRecipe(io.papermc.paper.inventory.@NotNull RecipeType<R, ?> recipeType, @Nullable ItemStack... items);
```

Comment:
Clarify the nullability of the varargs here.

(I'm not 100% sure how it works with varargs but I think what you've got now means `items` can be null) which isn't what you want, you want an element in items to be nullable)